### PR TITLE
[SPARK-31431][SQL] Add CalendarInterval encoder support

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

CalenderInterval is available to be converted to/from internal Spark SQL representation when it is a member of a Scala's product type e.g tuples/ case class etc but not as a primitive type

In this PR, we add a new ser/de rules to map CalenderInterval to CalenderIntervalType directly.

### Why are the changes needed?

improve CalenderInterval

### Does this PR introduce any user-facing change?

yes, CalenderInterval are now can be implicitly converted to Dataset/DataFame without being wrapped as scala Products
### How was this patch tested?

add new tests

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/Encoders.scala [hunk 1]
```diff
@@ -135,6 +136,13 @@ object Encoders {
    */
   def BINARY: Encoder[Array[Byte]] = ExpressionEncoder()
 
+  /**
+   * An encoder for [[CalendarInterval]]
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/Encoders.scala:140]: For some reasons, looks we failed to link `CalendarInterval`. We can make it `` `CalendarInterval` ``, for now.

```
[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/catalyst/target/java/org/apache/spark/sql/Encoders.java:109: error: reference not found
[error]    * An encoder for {@link CalendarInterval}
[error]                            ^
```

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/Encoders.scala:140]: ah, thanks

**Comment by yaooqinn** [comment]
cc @dongjoon-hyun @cloud-fan thanks for reviewing

File: sql/catalyst/src/main/scala/org/apache/spark/sql/Encoders.scala [hunk 2]
```diff
@@ -135,6 +136,13 @@ object Encoders {
    */
   def BINARY: Encoder[Array[Byte]] = ExpressionEncoder()
 
+  /**
+   * An encoder for `CalendarInterval`
+   *
+   * @since 3.1.0
+   */
+  def CALENDARINTERVAL: Encoder[CalendarInterval] = ExpressionEncoder()
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/Encoders.scala:144]: how about just interval?

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 3]
```diff
@@ -2358,6 +2358,18 @@ class DataFrameSuite extends QueryTest
     val df = Seq((1, new CalendarInterval(1, 2, 3))).toDF("a", "b")
     checkAnswer(df.selectExpr("b"), Row(new CalendarInterval(1, 2, 3)))
   }
+
+  test("CalendarInterval encoder support") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2362]: how about java bean?

Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2362]: the java bean support is added in this PR too

File: sql/catalyst/src/main/scala/org/apache/spark/sql/Encoders.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/Encoders.scala]: SGTM

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala [hunk 4]
```diff
@@ -237,6 +238,8 @@ abstract class SQLImplicits extends LowPrioritySQLImplicits {
    */
   implicit def symbolToColumn(s: Symbol): ColumnName = new ColumnName(s.name)
 
+  /** @since 3.1.0 */
+  implicit def newIntervalEncoder: Encoder[CalendarInterval] = Encoders.INTERVAL
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala:242]: nit: can we move it closer to where we define other encoders?

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java [hunk 5]
```diff
@@ -263,6 +264,70 @@ public void testSelect() {
...
+      ds.select(expr("value + interval 1 hour").cast("string")).as(Encoders.STRING());
+
+    Assert.assertEquals(
+      Arrays.asList("1 months 2 days 1 hours", "4 months 5 days 1 hours", null),
+      selected.collectAsList());
+  }
+
+
+  public static class IntervalBean implements Serializable {
+    private CalendarInterval i;
```
Comment (cloud-fan) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java:286]: can we test interval array as well?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala [hunk 6]
```diff
@@ -235,6 +235,7 @@ object ScalaReflection extends ScalaReflection {
       case t if isSubtype(t, localTypeOf[java.lang.String]) =>
         createDeserializerForString(path, returnNullable = false)
 
+      case t if isSubtype(t, localTypeOf[CalendarInterval]) => path
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:238]: just for curiosity: why interval inside array/map/struct works before this PR?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:238]: Do you mean we work for scala Products? `Seq(Array(something)).toDF` seems need that `something` to be java primitives or `scala.Product`s. Not yet support `CalendarInterval` or e.g....
```scala
scala> case class A(a: Int)
defined class A

scala> Seq(Array(A(1))).toDF
res8: org.apache.spark.sql.DataFrame = [value: array<struct<a:int>>]

scala> class B(a: Int)
defined class B

scala> Seq(Array(new B(1))).toDF
<console>:29: error: value toDF is not a member of Seq[Array[B]]
       Seq(Array(new B(1))).toDF
                            ^

scala> Seq(new java.math.BigDecimal(1)).toDF
res10: org.apache.spark.sql.DataFrame = [value: decimal(38,18)]

scala> Seq(Array(new java.math.BigDecimal(1))).toDF
<console>:28: error: value toDF is not a member of Seq[Array[java.math.BigDecimal]]
       Seq(Array(new java.math.BigDecimal(1))).toDF
```

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:238]: They don't work
```scala
scala> Seq(Map(new org.apache.spark.unsafe.types.CalendarInterval(1,2,3) -> new org.apache.spark.unsafe.types.CalendarInterval(1,2,3))).toDF
java.lang.UnsupportedOperationException: Schema for type org.apache.spark.unsafe.types.CalendarInterval is not supported
  at org.apache.spark.sql.catalyst.ScalaReflection$.$anonfun$schemaFor$1(ScalaReflection.scala:735)
  at scala.reflect.internal.tpe.TypeConstraints$UndoLog.undo(TypeConstraints.scala:69)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects(ScalaReflection.scala:877)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects$(ScalaReflection.scala:876)
  at org.apache.spark.sql.catalyst.ScalaReflection$.cleanUpReflectionObjects(ScalaReflection.scala:47)
  at org.apache.spark.sql.catalyst.ScalaReflection$.schemaFor(ScalaReflection.scala:663)
  at org.apache.spark.sql.catalyst.ScalaReflection$.$anonfun$schemaFor$1(ScalaReflection.scala:689)
  at scala.reflect.internal.tpe.TypeConstraints$UndoLog.undo(TypeConstraints.scala:69)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects(ScalaReflection.scala:877)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects$(ScalaReflection.scala:876)
  at org.apache.spark.sql.catalyst.ScalaReflection$.cleanUpReflectionObjects(ScalaReflection.scala:47)
  at org.apache.spark.sql.catalyst.ScalaReflection$.schemaFor(ScalaReflection.scala:663)
  at org.apache.spark.sql.catalyst.ScalaReflection$.deserializerForType(ScalaReflection.scala:164)
  at org.apache.spark.sql.catalyst.encoders.ExpressionEncoder$.apply(ExpressionEncoder.scala:55)
  at org.apache.spark.sql.SQLImplicits.newMapEncoder(SQLImplicits.scala:175)
  ... 49 elided

scala> Seq(Seq(new org.apache.spark.unsafe.types.CalendarInterval(1,2,3)).toDF
     | ;
<console>:2: error: ')' expected but ';' found.
       ;
       ^

scala> Seq(Seq(new org.apache.spark.unsafe.types.CalendarInterval(1,2,3))).toDF
java.lang.UnsupportedOperationException: Schema for type org.apache.spark.unsafe.types.CalendarInterval is not supported
  at org.apache.spark.sql.catalyst.ScalaReflection$.$anonfun$schemaFor$1(ScalaReflection.scala:735)
  at scala.reflect.internal.tpe.TypeConstraints$UndoLog.undo(TypeConstraints.scala:69)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects(ScalaReflection.scala:877)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects$(ScalaReflection.scala:876)
  at org.apache.spark.sql.catalyst.ScalaReflection$.cleanUpReflectionObjects(ScalaReflection.scala:47)
  at org.apache.spark.sql.catalyst.ScalaReflection$.schemaFor(ScalaReflection.scala:663)
  at org.apache.spark.sql.catalyst.ScalaReflection$.toCatalystArray$1(ScalaReflection.scala:430)
  at org.apache.spark.sql.catalyst.ScalaReflection$.$anonfun$serializerFor$1(ScalaReflection.scala:448)
  at scala.reflect.internal.tpe.TypeConstraints$UndoLog.undo(TypeConstraints.scala:69)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects(ScalaReflection.scala:877)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects$(ScalaReflection.scala:876)
  at org.apache.spark.sql.catalyst.ScalaReflection$.cleanUpReflectionObjects(ScalaReflection.scala:47)
  at org.apache.spark.sql.catalyst.ScalaReflection$.serializerFor(ScalaReflection.scala:410)
  at org.apache.spark.sql.catalyst.ScalaReflection$.$anonfun$serializerForType$1(ScalaReflection.scala:399)
  at scala.reflect.internal.tpe.TypeConstraints$UndoLog.undo(TypeConstraints.scala:69)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects(ScalaReflection.scala:877)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects$(ScalaReflection.scala:876)
  at org.apache.spark.sql.catalyst.ScalaReflection$.cleanUpReflectionObjects(ScalaReflection.scala:47)
  at org.apache.spark.sql.catalyst.ScalaReflection$.serializerForType(ScalaReflection.scala:391)
  at org.apache.spark.sql.catalyst.encoders.ExpressionEncoder$.apply(ExpressionEncoder.scala:54)
  at org.apache.spark.sql.SQLImplicits.newSequenceEncoder(SQLImplicits.scala:171)
  ... 49 elided

scala> Seq(Array(new org.apache.spark.unsafe.types.CalendarInterval(1,2,3))).toDF
<console>:28: error: value toDF is not a member of Seq[Array[org.apache.spark.unsafe.types.CalendarInterval]]
       Seq(Array(new org.apache.spark.unsafe.types.CalendarInterval(1,2,3))).toDF
```

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:239]: Sorry, I mean, struct:
> CalenderInterval is available to be converted to/from internal Spark SQL representation when it is a member of a Scala's product type e.g tuples/ case class etc but not as a primitive type

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:239]: with PR https://github.com/apache/spark/pull/28198/files  and the existing logic in `serializerFor`
1. https://github.com/yaooqinn/spark/blob/6cd0bef7fe872b5d9472a2140db188fb9f1eede8/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala#L436
2. https://github.com/yaooqinn/spark/blob/6cd0bef7fe872b5d9472a2140db188fb9f1eede8/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala#L554

**Comment by yaooqinn** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala [hunk 7]
```diff
@@ -126,6 +126,7 @@ object ScalaReflection extends ScalaReflection {
       case t if isSubtype(t, definitions.ShortTpe) => classOf[Array[Short]]
       case t if isSubtype(t, definitions.ByteTpe) => classOf[Array[Byte]]
       case t if isSubtype(t, definitions.BooleanTpe) => classOf[Array[Boolean]]
+      case t if isSubtype(t, localTypeOf[CalendarInterval]) => classOf[Array[CalendarInterval]]
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:129]: Is it really needed? `Decimal` is not handled here.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala [hunk 8]
```diff
@@ -210,6 +214,9 @@ abstract class SQLImplicits extends LowPrioritySQLImplicits {
   /** @since 1.6.1 */
   implicit def newStringArrayEncoder: Encoder[Array[String]] = ExpressionEncoder()
 
+  /** @since 3.1.0 */
+  implicit def newIntervalArrayEncoder: Encoder[Array[CalendarInterval]] = ExpressionEncoder()
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala:218]: ditto, why is there no decimal array encoder?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala (hunk 7 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:129]: Decimal is not supported，I will make it work in another PR

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java [hunk 9]
```diff
@@ -263,6 +264,80 @@ public void testSelect() {
...
+    private CalendarInterval[] ia;
```
Comment (cloud-fan) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java:287]: let's test more: map, list, another java bean.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 10]
```diff
@@ -2358,6 +2358,20 @@ class DataFrameSuite extends QueryTest
     val df = Seq((1, new CalendarInterval(1, 2, 3))).toDF("a", "b")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2358]: can we merge this test into the new one? they are testing the same thing.

Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2358]: OK

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java (hunk 9 above)
Comment (yaooqinn) [reply sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java:287]: updated

**Review by viirya — COMMENTED** [review]
Why do we need to add CalendarInterval Encoder support? It is Spark internal type used in row schema.

**Comment by yaooqinn** [comment]
It will be public after 3.0.0 release @viirya

**Comment by viirya** [comment]
That is not what I meant. My point is adding encoder support is for domain objects users frequently use to encode/decode to/from Spark internals. Do we expect users to use CalendarInterval in their domain objects?

**Comment by yaooqinn** [comment]
Hi, take your commit as an example https://github.com/apache/spark/commit/48e44b24a7663142176102ac4c6bf4242f103804, `Seq(Set(interval)).toDF()`, do intervals work as domain objects already?

**Comment by viirya** [comment]
Do we expect users to read data and represent as CalendarInterval in Dataset? Seems to me CalendarInterval is only for usage in Spark row. Although not the same, it sounds similar to me to make UTF8String in Dataset.

For domain objects we provide encoders in Dataset, I think they shall be frequently used domain objects which users will use in their business logic.

This sounds a strange encoder when I looked at it at first. It seems not to be a problem, however I wonder we should be careful in adding encoder. If others also agree to add, I'm fine for this.

**Comment by cloud-fan** [comment]
I'd say `CalendarInterval` should be treated the same as `Decimal`. They are semi-public, and are already supported partially (inside case class). It's arguable if we want to support more.

**Comment by yaooqinn** [comment]
For timestamp and date type, we already add `Instant` and `LocalDate` from java8 time API. In that case, we are very likely to add `Duration` and `Period` later for CalendarIntervalType too, then it seems more reasonable to add `CalendarInterval` encoder support as the encoder has half-implemented already.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/Encoders.scala [hunk 11]
```diff
@@ -27,6 +27,7 @@ import org.apache.spark.sql.catalyst.encoders.{encoderFor, ExpressionEncoder}
 import org.apache.spark.sql.catalyst.expressions.{BoundReference, Cast}
 import org.apache.spark.sql.catalyst.expressions.objects.{DecodeUsingSerializer, EncodeUsingSerializer}
 import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.CalendarInterval
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/Encoders.scala:30]: Why don't we move `CalendarInterval` to `org.apache.spark.types` before we more publicly expose this?

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/Encoders.scala:30]: If we're not sure whether to make this as a proper public type or not, let's don't add it for now.

**Pull request closed without merging**